File: examples/test-game/scenes/settings_menu/settings_menu.c
#include "settings_menu.h"

float x = 20, y = 20;
float speed = 2.0f;

EXPORT void Init(Engine_t *engine) {
    // Initialization logic for the main menu
}

EXPORT void ProcessInput(Engine_t *engine) {
    if (IsKeyDown(KEY_W)) y -= speed;
    if (IsKeyDown(KEY_S)) y += speed;
    if (IsKeyDown(KEY_A)) x -= speed;
    if (IsKeyDown(KEY_D)) x += speed;
}

EXPORT void Update(Engine_t *engine) {
    // Update logic for the main menu
    printf("Settings menu game update!\n");
}

EXPORT void Draw(Engine_t *engine) {
    // Drawing logic for the main menu
    DrawRectangle(x, y, 20, 20, BLUE);
}

EXPORT void Cleanup(Engine_t *engine) {
    // Cleanup logic for the main menu
    Engine_RunHook(engine, "HOOK_BEFORE_SETTINGS_MENU_CLEANUP");
    Engine_RunHook(engine, "HOOK_AFTER_SETTINGS_MENU_CLEANUP");
}

File: examples/holojam/mods/core-mod/settings_menu.lua
local player_speed = 4.0

local BLUE = { r = 50, g = 41, b = 55, a = 255 }

RegisterFunction("HOOK_SETTINGS_MENU_INIT", function()
  print("(Core Mod): Settings menu init!")

  player_x = 20
  player_y = 20
end)

RegisterFunction("HOOK_SETTINGS_MENU_PROCESS_INPUT", function()
  -- print("(Core Mod): Settings menu process input!")
  if IsKeyDown("KEY_F") then
    event_fire("TEST", "settings_menu_sender")
  end
end)

RegisterFunction("HOOK_SETTINGS_MENU_UPDATE", function()
  if IsKeyDown("KEY_W") then
    player_y = player_y - player_speed
  end
  if IsKeyDown("KEY_S") then
    player_y = player_y + player_speed
  end
  if IsKeyDown("KEY_A") then
    player_x = player_x - player_speed
  end
  if IsKeyDown("KEY_D") then
    player_x = player_x + player_speed
  end
end)

RegisterFunction("HOOK_SETTINGS_MENU_DRAW", function()
  DrawRectangle(player_x, player_y, 20, 20, BLUE)
end)

RegisterFunction("HOOK_SETTINGS_MENU_CLEANUP", function()
  print("(Core Mod): Settings menu cleanup!")
end)
